Libs/jack-audio-connection-kit-0.103.0/jackd/jackd.1
.TH "JACKD" "1" "0.103.0" "April 2006" ""
.SH "NAME"
jackd, jackstart \- JACK Audio Connection Kit sound server
.SH "SYNOPSYS"
\fBjackd\fR [\fIoptions\fR] \fB\-d\fI backend \fR 
[\fIbackend\-parameters\fR]
.br
\fBjackstart\fR [\fIoptions\fR] \fB\-d\fI backend \fR 
[\fIbackend\-parameters\fR]
.br
\fBjackd \-\-help\fR
.SH "DESCRIPTION"
\fBjackd\fR invokes the JACK audio server daemon.  The \fBjackstart\fR
command provides JACK's built\-in support for Linux 2.4.x kernels with
the realtime capabilities patch.  Use \fBjackd\fR for Linux 2.6.x and
all other systems.  All parameters are the same.

JACK is a low\-latency audio server.  Originally written for the
GNU/Linux operating system, it also supports Mac OS X and various Unix
platforms.  JACK can connect a number of different client applications
to an audio device and also to each other.  Most clients are external,
running in their own processes as normal applications.  JACK also
supports internal clients, which run within the \fBjackd\fR process
using a loadable "plugin" interface.

JACK differs from other audio servers in being designed from the
ground up for professional audio work.  It focuses on two key areas:
synchronous execution of all clients, and low latency operation.

For the latest JACK information, please consult the web site,
<\fBhttp://www.jackaudio.org\fR>.
.SH "OPTIONS"
.TP
\fB\-d, \-\-driver \fIbackend\fR [\fIbackend\-parameters\fR ]
.br
Select the audio interface backend.  The current list of supported
backends is: \fBalsa\fR, \fBcoreaudio\fR, \fBdummy\fR, \fBfreebob\fR,
\fBoss\fR and \fBportaudio\fR.  They are not all available on all
platforms.  All \fIbackend\-parameters\fR are optional.

.TP
\fB\-h, \-\-help\fR
.br
Print a brief usage message describing the main \fBjackd\fR options.
These do not include \fIbackend\-parameters\fR, which are listed using
the \fB\-\-help\fR option for each specific backend.  Examples below
show how to list them.
.TP
\fB\-m, \-\-no\-mlock\fR
Do not attempt to lock memory, even if \fB\-\-realtime\fR.
.TP
\fB\-n, \-\-name\fR \fIserver\-name\fR
Name this \fBjackd\fR instance \fIserver\-name\fR.  If unspecified,
this name comes from the \fB$JACK_DEFAULT_SERVER\fR environment
variable.  It will be "default" if that is not defined.
.TP
\fB\-p, \-\-port\-max \fI n\fR
Set the maximum number of ports the JACK server can manage.  
The default value is 256.
.TP
\fB\-R, \-\-realtime\fR 
.br
Use realtime scheduling.  This is needed for reliable low\-latency
performance.  On most systems, it requires \fBjackd\fR to run with
special scheduler and memory allocation privileges, which may be
obtained in several ways.  The simplest, and least-secure method is to
run \fBjackd\fR with root privileges.  This means that all JACK
clients must also run as root.  With a Linux 2.6 kernel, ordinary
users can run \fBjackd\fR and its clients using options of the
realtime LSM.  Linux 2.4 kernels need "POSIX draft capabilities"
enabled (see the \fB<linux/capability.h>\fR include file).  Using that
method, ordinary users who invoke the daemon using \fBjackstart\fR,
can later launch JACK clients without running them as root.  See
.I http://jackit.sourceforge.net/docs/faq.php#a52 
for more information.
.TP
\fB\-P, \-\-realtime\-priority \fIint\fR
When running \fB\-\-realtime\fR, set the scheduler priority to
\fIint\fR.
.TP
\fB\-\-silent\fR
Silence any output during operation.
.TP
\fB\-T, \-\-temporary\fR
Exit once all clients have closed their connections.
.TP
\fB\-t, \-\-timeout \fIint\fR
.br
Set client timeout limit in milliseconds.  The default is 500 msec.
In realtime mode the client timeout must be smaller than the watchdog timeout (5000 msec).
.TP
\fB\-u, \-\-unlock\fR
.br
Unlock libraries GTK+, QT, FLTK, Wine.
.TP
\fB\-v, \-\-verbose\fR
Give verbose output.
.TP
\fB\-c, \-\-clocksource\fR (\fI c(ycle)\fR | \fI h(pet) \fR | \fI s(ystem) \fR)
Select a specific wall clock (Cycle Counter, HPET timer, System timer).
.TP
\fB\-V, \-\-version\fR
Print the current JACK version number and exit.
.SS ALSA BACKEND OPTIONS
.TP
\fB\-C, \-\-capture\fR [ \fIname\fR ]
Provide only capture ports, unless combined with \-D or \-P.  Parameterally set 
capture device name.
.TP
\fB\-d, \-\-device \fIname\fR
.br
The ALSA pcm device \fIname\fR to use.  If none is specified, JACK will
use "hw:0", the first hardware card defined in \fB/etc/modules.conf\fR.
.TP
\fB\-z, \-\-dither [rectangular,triangular,shaped,none]
Set dithering mode.  If \fBnone\fR or unspecified, dithering is off.
Only the first letter of the mode name is required.
.TP
\fB\-D, \-\-duplex\fR
Provide both capture and playback ports.  Defaults to on unless only one 
of \-P or \-C is specified.
.TP
\fB\-h, \-\-help\fR Print a brief usage message describing only the
\fBalsa\fR backend parameters.
.TP
\fB\-M, \-\-hwmeter\fR
.br
Enable hardware metering for devices that support it.  Otherwise, use
software metering.
.TP
\fB\-H, \-\-hwmon\fR
.br
Enable hardware monitoring of capture ports.  This is a method for
obtaining "zero latency" monitoring of audio input.  It requires
support in hardware and from the underlying ALSA device driver.

When enabled, requests to monitor capture ports will be satisfied by
creating a direct signal path between audio interface input and output
connectors, with no processing by the host computer at all.  This
offers the lowest possible latency for the monitored signal.

Presently (March 2003), only the RME Hammerfall series and cards based
on the ICE1712 chipset (M\-Audio Delta series, Terratec, and others)
support \fB\-\-hwmon\fR.  In the future, some consumer cards may also
be supported by modifying their mixer settings.

Without \fB\-\-hwmon\fR, port monitoring requires JACK to read audio
into system memory, then copy it back out to the hardware again,
imposing the basic JACK system latency determined by the
\fB\-\-period\fR and \fB\-\-nperiods\fR parameters.
.TP
\fB\-i, \-\-inchannels \fIint\fR
.br
Number of capture channels.  Default is maximum supported by hardware.
.TP
\fB\-n, \-\-nperiods \fIint\fR
.br
Specify the number of periods of playback latency.  In seconds, this
corresponds to \fB\-\-nperiods\fR times \fB\-\-period\fR divided by
\fB\-\-rate\fR.  The default is 2, the minimum allowable.  For most
devices, there is no need for any other value with the
\fB\-\-realtime\fR option.  Without realtime privileges or with boards
providing unreliable interrupts (like ymfpci), a larger value may
yield fewer xruns.  This can also help if the system is not tuned for
reliable realtime scheduling.

For most ALSA devices, the hardware buffer has exactly
\fB\-\-period\fR times \fB\-\-nperiods\fR frames.  Some devices demand
a larger buffer.  If so, JACK will use the smallest possible buffer
containing at least \fB\-\-nperiods\fR, but the playback latency does
not increase.
.TP
\fB\-o, \-\-outchannels \fIint\fR
.br
Number of playback channels.  Default is maximum supported by hardware.
.TP
\fB\-P, \-\-playback\fR [ \fIname\fR ]
Provide only playback ports, unless combined with \-D or \-C.  Optionally set 
playback device name.
.TP
\fB\-p, \-\-period \fIint\fR
.br
Specify the number of frames between JACK \fBprocess()\fR calls.  The
default is 1024.  If you need low latency, set \fB\-p\fR as low as you
can go without seeing xruns.  A larger period size yields higher
latency, but makes xruns less likely.  The JACK capture latency in
seconds is \fB\-\-period\fR divided by \fB\-\-rate\fR.
.TP
\fB\-r, \-\-rate \fIint\fR
Specify the sample rate.  The default is 48000.
.TP
\fB\-S, \-\-shorts
.br
Try to configure card for 16\-bit samples first, only trying 32\-bits if
unsuccessful.  Default is to prefer 32\-bit samples.
.TP
\fB\-s, \-\-softmode\fR 
.br
Ignore xruns reported by the ALSA driver.  This makes JACK less likely
to disconnect unresponsive ports when running without \fB\-\-realtime\fR.
.SS COREAUDIO BACKEND PARAMETERS
.TP
\fB\-c \-\-channel\fR
Maximum number of channels (default: 2)
.TP
\fB\-i \-\-channelin\fR
Maximum number of input channels (default: 2)
.TP
\fB\-o \-\-channelout\fR
Maximum number of output channels (default: 2)
.TP
\fB\-C \-\-capture\fR
Whether or not to capture (default: true)
.TP
\fB\-P \-\-playback\fR
Whether or not to playback (default: true)
.TP
\fB\-D \-\-duplex\fR
Capture and playback (default: true)
.TP
\fB\-r \-\-rate\fR
Sample rate (default: 44100)
.TP
\fB\-p \-\-period\fR
Frames per period (default: 128)
.TP
\fB\-n \-\-name\fR
Driver name (default: none)
.TP
\fB\-I \-\-id\fR
Audio Device ID (default: 0)
.SS DUMMY BACKEND PARAMETERS
.TP
\fB\-C, \-\-capture \fIint\fR
Specify number of capture ports. The default value is 2.
.TP
\fB\-P, \-\-playback \fIint\fR
Specify number of playback ports. The default value is 2.
.TP
\fB\-r, \-\-rate \fIint\fR
Specify sample rate. The default value is 48000.
.TP
\fB\-p, \-\-period \fIint\fR
Specify number of frames per period. The default value is 1024.
If you need low latency, set \fB\-p\fR as low as you
can go without seeing xruns.  A larger period size yields higher
latency, but makes xruns less likely.  The JACK capture latency in
seconds is \fB\-\-period\fR divided by \fB\-\-rate\fR.
.TP
\fB\-w, \-\-wait \fIint\fR 
Specify number of usecs to wait between engine processes. 
The default value is 21333.
.SS OSS BACKEND PARAMETERS
.TP
\fB\-r, \-\-rate \fIint\fR
Specify the sample rate.  The default is 48000.
.TP
\fB\-p, \-\-period \fIint\fR
Specify number of frames between JACK \fBprocess()\fR calls.  The
default is 1024.  If you need low latency, set \fB\-p\fR as low as you
can go without seeing xruns.  A larger period size yields higher
latency, but makes xruns less likely.  The JACK capture latency in
seconds is \fB\-\-period\fR divided by \fB\-\-rate\fR.
.TP
\fB\-n, \-\-nperiods \fIint\fR
Specify the number of periods in the hardware buffer.  The default is
2.  The period size (\fB\-p\fR) times \fB\-\-nperiods\fR times four is
the JACK buffer size in bytes.  The JACK output latency in seconds is
\fB\-\-nperiods\fR times \fB\-\-period\fR divided by \fB\-\-rate\fR.
.TP
\fB\-w, \-\-wordlength \fIint\fR
Specify the sample size in bits. The default is 16.
.TP
\fB\-i, \-\-inchannels \fIint\fR
Specify how many channels to capture (default: 2)
.TP
\fB\-o, \-\-outchannels \fIint\fR
Specify number of playback channels (default: 2)
.TP
\fB\-C, \-\-capture \fIdevice_file\fR
Specify input device for capture (default: /dev/dsp)
.TP
\fB\-P, \-\-playback \fIdevice_file\fR
Specify output device for playback (default: /dev/dsp)
.TP
\fB\-b, \-\-ignorehwbuf \fIboolean\fR
Specify, whether to ignore hardware period size (default: false)
.SS PORTAUDIO BACKEND PARAMETERS
.TP
\fB\-c \-\-channel\fR
Maximum number of channels (default: all available hardware channels)
.TP
\fB\-i \-\-channelin\fR
Maximum number of input channels (default: all available hardware channels)
.TP
\fB\-o \-\-channelout\fR
Maximum number of output channels (default: all available hardware channels)
.TP
\fB\-C \-\-capture\fR
Whether or not to capture (default: true)
.TP
\fB\-P \-\-playback\fR
Whether or not to playback (default: true)
.TP
\fB\-D \-\-duplex\fR
Capture and playback (default: true)
.TP
\fB\-r \-\-rate\fR
Sample rate (default: 48000)
.TP
\fB\-p \-\-period\fR
Frames per period (default: 1024)
.TP
\fB\-n \-\-name\fR
Driver name (default: none)
.TP
\fB\-z \-\-dither\fR
Dithering mode (default: none)
.SH "EXAMPLES"
.PP
Print usage message for the parameters specific to each backend.
.IP
\fBjackd \-d alsa \-\-help\fR
.br
\fBjackd \-d coreaudio \-\-help\fR
.br
\fBjackd \-d dummy \-\-help\fR
.br
\fBjackd \-d freebob \-\-help\fR
.br
\fBjackd \-d oss \-\-help\fR
.br
\fBjackd \-d portaudio \-\-help\fR
.PP
Run the JACK daemon with realtime priority using the first ALSA
hardware card defined in \fB/etc/modules.conf\fR.
.IP
\fBjackstart \-\-realtime \-\-driver=alsa\fR
.PP
Run the JACK daemon with low latency giving verbose output, which can
be helpful for trouble\-shooting system latency problems.  A
reasonably well\-tuned system with a good sound card and a
low\-latency kernel can handle these values reliably.  Some can do
better.  If you get xrun messages, try a larger buffer.  Tuning a
system for low latency can be challenging.  The JACK FAQ,
.I http://jackit.sourceforge.net/docs/faq.php\fR
has some useful suggestions.
.IP
\fBjackstart \-Rv \-d alsa \-p 128 \-n 2 \-r 44100\fR
.PP
Run \fBjackd\fR with realtime priority using the "sblive" ALSA device
defined in ~/.asoundrc.  Apply shaped dithering to playback audio.
.IP
\fBjackd \-R \-d alsa \-d sblive \-\-dither=shaped\fR
.PP
Run \fBjackd\fR with no special privileges using the second ALSA
hardware card defined in \fB/etc/modules.conf\fR.  Any xruns reported
by the ALSA backend will be ignored.  The larger buffer helps reduce
data loss.  Rectangular dithering will be used for playback.
.IP
\fBjackd \-d alsa \-d hw:1 \-p2048 \-n3 \-\-softmode \-zr\fR
.PP

Libs/jack-audio-connection-kit-0.103.0/jackd/jackd.c
/* -*- mode: c; c-file-style: "bsd"; -*- */
/*
    Copyright (C) 2001-2005 Paul Davis
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    $Id: jackd.c 1010 2007-02-23 19:23:07Z paul $
*/

#include <config.h>

#include <stdio.h>
#include <ctype.h>
#include <signal.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/shm.h>
#include <sys/wait.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>
#include <dirent.h>
#include <dlfcn.h>

#include <jack/engine.h>
#include <jack/internal.h>
#include <jack/driver.h>
#include <jack/shm.h>
#include <jack/driver_parse.h>
#include <jack/messagebuffer.h>

#ifdef USE_CAPABILITIES

#include <sys/stat.h>
/* capgetp and capsetp are linux only extensions, not posix */
#undef _POSIX_SOURCE
#include <sys/capability.h>
#include <jack/start.h>

static struct stat pipe_stat;

#endif /* USE_CAPABILITIES */

static JSList *drivers = NULL;
static sigset_t signals;
static jack_engine_t *engine = NULL;
static char *server_name = NULL;
static int realtime = 0;
static int realtime_priority = 10;
static int do_mlock = 1;
static int temporary = 0;
static int verbose = 0;
static int client_timeout = 0; /* msecs; if zero, use period size. */
static unsigned int port_max = 256;
static int do_unlock = 0;
static jack_nframes_t frame_time_offset = 0;

static void 
do_nothing_handler (int sig)
{
	/* this is used by the child (active) process, but it never
	   gets called unless we are already shutting down after
	   another signal.
	*/
	char buf[64];
	snprintf (buf, sizeof(buf),
		  "received signal %d during shutdown (ignored)\n", sig);
	write (1, buf, strlen (buf));
}

static int
jack_main (jack_driver_desc_t * driver_desc, JSList * driver_params)
{
	int sig;
	int i;
	sigset_t allsignals;
	struct sigaction action;
	int waiting;

	/* ensure that we are in our own process group so that
	   kill (SIG, -pgrp) does the right thing.
	*/

	setsid ();

	pthread_setcanceltype (PTHREAD_CANCEL_ASYNCHRONOUS, NULL);

	/* what's this for?

	   POSIX says that signals are delivered like this:

	   * if a thread has blocked that signal, it is not
	       a candidate to receive the signal.
           * of all threads not blocking the signal, pick
	       one at random, and deliver the signal.

           this means that a simple-minded multi-threaded program can
           expect to get POSIX signals delivered randomly to any one
           of its threads,

	   here, we block all signals that we think we might receive
	   and want to catch. all "child" threads will inherit this
	   setting. if we create a thread that calls sigwait() on the
	   same set of signals, implicitly unblocking all those
	   signals. any of those signals that are delivered to the
	   process will be delivered to that thread, and that thread
	   alone. this makes cleanup for a signal-driven exit much
	   easier, since we know which thread is doing it and more
	   importantly, we are free to call async-unsafe functions,
	   because the code is executing in normal thread context
	   after a return from sigwait().
	*/

	sigemptyset (&signals);
	sigaddset(&signals, SIGHUP);
	sigaddset(&signals, SIGINT);
	sigaddset(&signals, SIGQUIT);
	sigaddset(&signals, SIGPIPE);
	sigaddset(&signals, SIGTERM);
	sigaddset(&signals, SIGUSR1);
	sigaddset(&signals, SIGUSR2);

	/* all child threads will inherit this mask unless they
	 * explicitly reset it 
	 */

	pthread_sigmask (SIG_BLOCK, &signals, 0);

	if (!realtime && client_timeout == 0)
		client_timeout = 500; /* 0.5 sec; usable when non realtime. */

	/* get the engine/driver started */

	if ((engine = jack_engine_new (realtime, realtime_priority, 
				       do_mlock, do_unlock, server_name,
				       temporary, verbose, client_timeout,
				       port_max, getpid(), frame_time_offset, drivers)) == 0) {
		fprintf (stderr, "cannot create engine\n");
		return -1;
	}

	fprintf (stderr, "loading driver ..\n");
	
	if (jack_engine_load_driver (engine, driver_desc, driver_params)) {
		fprintf (stderr, "cannot load driver module %s\n",
			 driver_desc->name);
		goto error;
	}

	if (engine->driver->start (engine->driver) != 0) {
		jack_error ("cannot start driver");
		goto error;
	}

	/* install a do-nothing handler because otherwise pthreads
	   behaviour is undefined when we enter sigwait.
	*/

	sigfillset (&allsignals);
	action.sa_handler = do_nothing_handler;
	action.sa_mask = allsignals;
	action.sa_flags = SA_RESTART|SA_RESETHAND;

	for (i = 1; i < NSIG; i++) {
		if (sigismember (&signals, i)) {
			sigaction (i, &action, 0);
		} 
	}
	
	if (verbose) {
		fprintf (stderr, "%d waiting for signals\n", getpid());
	}

	waiting = TRUE;

	while (waiting) {
		sigwait (&signals, &sig);

		fprintf (stderr, "jack main caught signal %d\n", sig);
		
		switch (sig) {
		case SIGUSR1:
			jack_dump_configuration(engine, 1);
			break;
		case SIGUSR2:
			/* driver exit */
			waiting = FALSE;
			break;
		default:
			waiting = FALSE;
			break;
		}
	} 
	
	if (sig != SIGSEGV) {

		/* unblock signals so we can see them during shutdown.
		   this will help prod developers not to lose sight of
		   bugs that cause segfaults etc. during shutdown.
		*/
		sigprocmask (SIG_UNBLOCK, &signals, 0);
	}
	
	jack_engine_delete (engine);
	return 1;
	
error:
	jack_engine_delete (engine);
	return -1;
}

static jack_driver_desc_t *
jack_drivers_get_descriptor (JSList * drivers, const char * sofile)
{
	jack_driver_desc_t * descriptor, * other_descriptor;
	JackDriverDescFunction so_get_descriptor;
	JSList * node;
	void * dlhandle;
	char * filename;
	const char * dlerr;
	int err;
	char* driver_dir;

	if ((driver_dir = getenv("JACK_DRIVER_DIR")) == 0) {
		driver_dir = ADDON_DIR;
	}
	filename = malloc (strlen (driver_dir) + 1 + strlen (sofile) + 1);
	sprintf (filename, "%s/%s", driver_dir, sofile);

	if (verbose) {
		fprintf (stderr, "getting driver descriptor from %s\n", filename);
	}

	if ((dlhandle = dlopen (filename, RTLD_NOW|RTLD_GLOBAL)) == NULL) {
		jack_error ("could not open driver .so '%s': %s\n", filename, dlerror ());
		free (filename);
		return NULL;
	}

	so_get_descriptor = (JackDriverDescFunction)
		dlsym (dlhandle, "driver_get_descriptor");

	if ((dlerr = dlerror ()) != NULL) {
		fprintf(stderr, "%s\n", dlerr);
		dlclose (dlhandle);
		free (filename);
		return NULL;
	}

	if ((descriptor = so_get_descriptor ()) == NULL) {
		jack_error ("driver from '%s' returned NULL descriptor\n", filename);
		dlclose (dlhandle);
		free (filename);
		return NULL;
	}

	if ((err = dlclose (dlhandle)) != 0) {
		jack_error ("error closing driver .so '%s': %s\n", filename, dlerror ());
	}

	/* check it doesn't exist already */
	for (node = drivers; node; node = jack_slist_next (node)) {
		other_descriptor = (jack_driver_desc_t *) node->data;

		if (strcmp (descriptor->name, other_descriptor->name) == 0) {
			jack_error ("the drivers in '%s' and '%s' both have the name '%s'; using the first\n",
				    other_descriptor->file, filename, other_descriptor->name);
			/* FIXME: delete the descriptor */
			free (filename);
			return NULL;
		}
	}

	snprintf (descriptor->file, sizeof(descriptor->file), "%s", filename);
	free (filename);

	return descriptor;
}

static JSList *
jack_drivers_load ()
{
	struct dirent * dir_entry;
	DIR * dir_stream;
	const char * ptr;
	int err;
	JSList * driver_list = NULL;
	jack_driver_desc_t * desc;
	char* driver_dir;

	if ((driver_dir = getenv("JACK_DRIVER_DIR")) == 0) {
		driver_dir = ADDON_DIR;
	}

	/* search through the driver_dir and add get descriptors
	   from the .so files in it */
	dir_stream = opendir (driver_dir);
	if (!dir_stream) {
		jack_error ("could not open driver directory %s: %s\n",
			    driver_dir, strerror (errno));
		return NULL;
	}
  
	while ( (dir_entry = readdir (dir_stream)) ) {
		/* check the filename is of the right format */
		if (strncmp ("jack_", dir_entry->d_name, 5) != 0) {
			continue;
		}

		ptr = strrchr (dir_entry->d_name, '.');
		if (!ptr) {
			continue;
		}
		ptr++;
		if (strncmp ("so", ptr, 2) != 0) {
			continue;
		}

		desc = jack_drivers_get_descriptor (drivers, dir_entry->d_name);
		if (desc) {
			driver_list = jack_slist_append (driver_list, desc);
		}
	}

	err = closedir (dir_stream);
	if (err) {
		jack_error ("error closing driver directory %s: %s\n",
			    driver_dir, strerror (errno));
	}

	if (!driver_list) {
		jack_error ("could not find any drivers in %s!\n", driver_dir);
		return NULL;
	}

	return driver_list;
}

static void copyright (FILE* file)
{
	fprintf (file, "jackd " VERSION "\n"
"Copyright 2001-2005 Paul Davis and others.\n"
"jackd comes with ABSOLUTELY NO WARRANTY\n"
"This is free software, and you are welcome to redistribute it\n"
"under certain conditions; see the file COPYING for details\n\n");
}

static void usage (FILE *file) 
{
	copyright (file);
	fprintf (file, "\n"
"usage: jackd [ --realtime OR -R [ --realtime-priority OR -P priority ] ]\n"
"             [ --name OR -n server-name ]\n"
"             [ --no-mlock OR -m ]\n"
"             [ --unlock OR -u ]\n"
"             [ --timeout OR -t client-timeout-in-msecs ]\n"
"             [ --port-max OR -p maximum-number-of-ports]\n"
"             [ --debug-timer OR -D ]\n"
"             [ --verbose OR -v ]\n"
"             [ --clocksource OR -c [ c(ycle) | h(pet) | s(ystem) ]\n"
"             [ --silent OR -s ]\n"
"             [ --version OR -V ]\n"
"         -d backend [ ... backend args ... ]\n"
"             The backend can be `alsa', `coreaudio', `dummy',\n"
"                                `freebob', `oss' or `portaudio'.\n\n"
"       jackd -d backend --help\n"
"             to display options for each backend\n\n");
}	

static jack_driver_desc_t *
jack_find_driver_descriptor (const char * name)
{
	jack_driver_desc_t * desc = 0;
	JSList * node;

	for (node = drivers; node; node = jack_slist_next (node)) {
		desc = (jack_driver_desc_t *) node->data;

		if (strcmp (desc->name, name) != 0) {
			desc = NULL;
		} else {
			break;
		}
	}

	return desc;
}

static void
jack_cleanup_files (const char *server_name)
{
	DIR *dir;
	struct dirent *dirent;
	char dir_name[PATH_MAX+1] = "";
        jack_server_dir (server_name, dir_name);

	/* On termination, we remove all files that jackd creates so
	 * subsequent attempts to start jackd will not believe that an
	 * instance is already running.  If the server crashes or is
	 * terminated with SIGKILL, this is not possible.  So, cleanup
	 * is also attempted when jackd starts.
	 *
	 * There are several tricky issues.  First, the previous JACK
	 * server may have run for a different user ID, so its files
	 * may be inaccessible.  This is handled by using a separate
	 * JACK_TMP_DIR subdirectory for each user.  Second, there may
	 * be other servers running with different names.  Each gets
	 * its own subdirectory within the per-user directory.  The
	 * current process has already registered as `server_name', so
	 * we know there is no other server actively using that name.
	 */

	/* nothing to do if the server directory does not exist */
	if ((dir = opendir (dir_name)) == NULL) {
		return;
	}

	/* unlink all the files in this directory, they are mine */
	while ((dirent = readdir (dir)) != NULL) {

		char fullpath[PATH_MAX+1];

		if ((strcmp (dirent->d_name, ".") == 0)
		    || (strcmp (dirent->d_name, "..") == 0)) {
			continue;
		}

		snprintf (fullpath, sizeof (fullpath), "%s/%s",
			  dir_name, dirent->d_name);

		if (unlink (fullpath)) {
			jack_error ("cannot unlink `%s' (%s)", fullpath,
				    strerror (errno));
		}
	} 

	closedir (dir);

	/* now, delete the per-server subdirectory, itself */
	if (rmdir (dir_name)) {
 		jack_error ("cannot remove `%s' (%s)", dir_name,
			    strerror (errno));
	}

	/* finally, delete the per-user subdirectory, if empty */
	if (rmdir (jack_user_dir ())) {
		if (errno != ENOTEMPTY) {
			jack_error ("cannot remove `%s' (%s)",
				    jack_user_dir (), strerror (errno));
		}
	}
}

static void
maybe_use_capabilities ()
{
#ifdef USE_CAPABILITIES
	int status;

	/* check to see if there is a pipe in the right descriptor */
	if ((status = fstat (PIPE_WRITE_FD, &pipe_stat)) == 0 &&
	    S_ISFIFO(pipe_stat.st_mode)) {

		/* tell jackstart we are up and running */
  	        char c = 1;

	        if (write (PIPE_WRITE_FD, &c, 1) != 1) {
		        fprintf (stderr, "cannot write to jackstart sync "
				 "pipe %d (%s)\n", PIPE_WRITE_FD,
				 strerror (errno));
	        }

		if (close(PIPE_WRITE_FD) != 0) {
			fprintf(stderr,
				"jackd: error on startup pipe close: %s\n",
				strerror (errno));
		} else {
			/* wait for jackstart process to set our capabilities */
			if (wait (&status) == -1) {
				fprintf (stderr, "jackd: wait for startup "
					 "process exit failed\n");
			}
			if (!WIFEXITED (status) || WEXITSTATUS (status)) {
				fprintf(stderr, "jackd: jackstart did not "
					"exit cleanly\n");
				exit (1);
			}
		}
	}
#endif /* USE_CAPABILITIES */
}

int	       
main (int argc, char *argv[])

{
	jack_driver_desc_t * desc;
	const char *options = "-ad:P:uvshVRTFlt:mn:p:c:";
	struct option long_options[] = 
	{ 
		{ "driver", 1, 0, 'd' },
		{ "verbose", 0, 0, 'v' },
		{ "help", 0, 0, 'h' },
		{ "tmpdir-location", 0, 0, 'l' },
		{ "port-max", 1, 0, 'p' },
		{ "no-mlock", 0, 0, 'm' },
		{ "name", 1, 0, 'n' },
		{ "unlock", 0, 0, 'u' },
		{ "realtime", 0, 0, 'R' },
		{ "realtime-priority", 1, 0, 'P' },
		{ "timeout", 1, 0, 't' },
		{ "temporary", 0, 0, 'T' },
		{ "version", 0, 0, 'V' },
		{ "silent", 0, 0, 's' },
		{ "clock-source", 1, 0, 'c' },
		{ 0, 0, 0, 0 }
	};
	int opt = 0;
	int option_index = 0;
	int seen_driver = 0;
	char *driver_name = NULL;
	char **driver_args = NULL;
	JSList * driver_params;
	int driver_nargs = 1;
	int show_version = 0;
	int i;
	int rc;

	setvbuf (stdout, NULL, _IOLBF, 0);

	maybe_use_capabilities ();

	opterr = 0;
	while (!seen_driver &&
	       (opt = getopt_long (argc, argv, options,
				   long_options, &option_index)) != EOF) {
		switch (opt) {

		case 'c':
			if (tolower (optarg[0]) == 'h') {
				clock_source = JACK_TIMER_HPET;
			} else if (tolower (optarg[0]) == 'c') {
				clock_source = JACK_TIMER_CYCLE_COUNTER;
			} else if (tolower (optarg[0]) == 's') {
				clock_source = JACK_TIMER_SYSTEM_CLOCK;
			} else {
				usage (stderr);
				return -1;
			}
			break;

		case 'd':
			seen_driver = 1;
			driver_name = optarg;
			break;

		case 'D':
			frame_time_offset = JACK_MAX_FRAMES - atoi(optarg); 
			break;

		case 'l':
			/* special flag to allow libjack to determine jackd's idea of where tmpdir is */
			printf ("%s\n", jack_tmpdir);
			exit (0);

		case 'm':
			do_mlock = 0;
			break;

		case 'n':
			server_name = optarg;
			break;

		case 'p':
			port_max = (unsigned int) atol (optarg);
			break;

		case 'P':
			realtime_priority = atoi (optarg);
			break;

		case 'R':
			realtime = 1;
			break;

		case 's':
			jack_set_error_function (silent_jack_error_callback);
			break;

		case 'T':
			temporary = 1;
			break;

		case 't':
			client_timeout = atoi (optarg);
			break;

		case 'u':
			do_unlock = 1;
			break;

		case 'v':
			verbose = 1;
			break;

		case 'V':
			show_version = 1;
			break;

		default:
			fprintf (stderr, "Unknown option character %c\n",
				 optopt);
			/*fallthru*/
		case 'h':
			usage (stdout);
			return -1;
		}
	}

	if (show_version) {
		printf ( "jackd version " VERSION 
				" tmpdir " DEFAULT_TMP_DIR 
				" protocol " PROTOCOL_VERSION
				"\n");
		return -1;
	}

	if (realtime && (client_timeout >= JACKD_WATCHDOG_TIMEOUT)) {
		usage (stderr);
		fprintf (stderr, "In realtime mode (-R) the client timeout must be smaller than the watchdog timeout (%ims).\n", JACKD_WATCHDOG_TIMEOUT);
		exit (1);
	}

	if (!seen_driver) {
		usage (stderr);
		exit (1);
	}

	drivers = jack_drivers_load ();
	if (!drivers) {
		fprintf (stderr, "jackd: no drivers found; exiting\n");
		exit (1);
	}

	desc = jack_find_driver_descriptor (driver_name);
	if (!desc) {
		fprintf (stderr, "jackd: unknown driver '%s'\n", driver_name);
		exit (1);
	}

	if (optind < argc) {
		driver_nargs = 1 + argc - optind;
	} else {
		driver_nargs = 1;
	}

	if (driver_nargs == 0) {
		fprintf (stderr, "No driver specified ... hmm. JACK won't do"
			 " anything when run like this.\n");
		return -1;
	}

	driver_args = (char **) malloc (sizeof (char *) * driver_nargs);
	driver_args[0] = driver_name;
	
	for (i = 1; i < driver_nargs; i++) {
		driver_args[i] = argv[optind++];
	}

	if (jack_parse_driver_params (desc, driver_nargs,
				      driver_args, &driver_params)) {
		exit (0);
	}

	if (server_name == NULL)
		server_name = jack_default_server_name ();

	copyright (stdout);

	rc = jack_register_server (server_name);
	switch (rc) {
	case EEXIST:
		fprintf (stderr, "`%s' server already active\n", server_name);
		exit (1);
	case ENOSPC:
		fprintf (stderr, "too many servers already active\n");
		exit (2);
	case ENOMEM:
		fprintf (stderr, "no access to shm registry\n");
		exit (3);
	default:
		if (verbose)
			fprintf (stderr, "server `%s' registered\n",
				 server_name);
	}

	/* clean up shared memory and files from any previous
	 * instance of this server name */
	jack_cleanup_shm ();
	jack_cleanup_files (server_name);

	/* run the server engine until it terminates */
	jack_main (desc, driver_params);

	/* clean up shared memory and files from this server instance */
	if (verbose)
		fprintf (stderr, "cleaning up shared memory\n");
	jack_cleanup_shm ();
	if (verbose)
		fprintf (stderr, "cleaning up files\n");
	jack_cleanup_files (server_name);
	if (verbose)
		fprintf (stderr, "unregistering server `%s'\n", server_name);
	jack_unregister_server (server_name);

	exit (0);
}
